pkg/p2p/peer/peer.go

```diff
@@ -180,31 +180,45 @@ func (p *Reader) Accept() error {
 	return nil
 }
 
-// Create two-way communication with a peer. This function serves to split up
-// the passed context and allowing either goroutine to cancel the other
-// in case of errors.
+// Create two-way communication with a peer. This function will allow both
+// goroutines to run as long as no errors are encountered. Once the first error
+// comes through, the context is canceled, and both goroutines are cleaned up.
 func Create(ctx context.Context, reader *Reader, writer *Writer, writeQueueChan <-chan bytes.Buffer) {
-	readCtx, cancelReader := context.WithCancel(ctx)
-	writeCtx, cancelWriter := context.WithCancel(ctx)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	go reader.ReadLoop(readCtx, cancelWriter)
-	go writer.Serve(writeCtx, cancelReader, writeQueueChan)
+	errChan := make(chan error, 1)
+
+	go reader.ReadLoop(ctx, errChan)
+	go writer.Serve(ctx, writeQueueChan, errChan)
+
+	// Errors are already getting logged with the proper fields attached.
+	// So, we will just wait for one error to come through, after which we can
+	// cancel the context and exit this function.
+	<-errChan
```

Comment:
The point of `errChan` is to handle errors (among which disconnections). This should be the responsibility of the caller (`Create`) to do. In some cases, we might have some errors that do not grant a disconnection (so a `cancel`). In any case, this is a bit of a nitpick and I will record it in a separate issue